### PR TITLE
Limit reports generation per customer

### PR DESCRIPTION
**Pull request by MatanYadaev**
CleanShot 2024-07-30 at 12 34 57

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling in data fetching, providing more informative feedback for users.
	- Specific error handling for business report creation with tailored notifications for error cases.

- **Improvements**
	- Added validation for the maximum number of business reports a customer can create, ensuring compliance with business rules.
	- Expanded customer data retrieval to include configuration details.

- **Schema Updates**
	- Introduced a new optional property in the configuration schema to manage limits on business reports.

- **Refactor**
	- Streamlined imports in the filter controller, removing unnecessary dependencies to improve maintainability.
	- Enhanced test setup by incorporating SecretsManagerFactory across workflow-related tests.

### CODE REVIEW

**Review by liorzam — APPROVED** [review]

**Comment by liorzam** [comment]
Please make lint and CI

File: apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts [hunk 1]
```diff
@@ -44,9 +44,11 @@ export const fetcher: IFetcher = async ({
...
-        ? json?.errors?.map(({ message }) => `${message}\n`)?.join('')
-        : message;
+      if (Array.isArray(json?.errors)) {
+        message = json?.errors?.map(({ message }) => `${message}\n`)?.join('');
+      } else if (json.message) {
+        message = json.message;
+      }
     }
 
     console.error(message);
```
Comment (tomer-shvadron) [inline apps/backoffice-v2/src/common/utils/fetcher/fetcher.ts:54]: Can we remove this `console.error`? @Omri-Levy

File: apps/backoffice-v2/src/domains/business-reports/hooks/mutations/useCreateBusinessReportMutation/useCreateBusinessReportMutation.tsx [hunk 2]
```diff
@@ -51,6 +52,12 @@ export const useCreateBusinessReportMutation = ({
       onSuccess?.(data);
     },
     onError: error => {
+      if (error instanceof HttpError && error.code === 400) {
```
Comment (tomer-shvadron) [inline apps/backoffice-v2/src/domains/business-reports/hooks/mutations/useCreateBusinessReportMutation/useCreateBusinessReportMutation.tsx:55]: Should we use [HTTP error code 429](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)?

File: services/workflows-service/src/business-report/business-report.controller.internal.ts [hunk 3]
```diff
@@ -55,6 +56,26 @@ export class BusinessReportControllerInternal {
     }: CreateBusinessReportDto,
     @CurrentProject() currentProjectId: TProjectId,
   ) {
+    const customer = await this.customerService.getByProjectId(currentProjectId);
+    const maxBusinessReports = customer.config?.maxBusinessReports;
+
+    if (isNumber(maxBusinessReports) && maxBusinessReports > 0) {
+      const businessReportsCount = await this.businessReportService.count(
+        {
+          where: {
```
Comment (tomer-shvadron) [inline services/workflows-service/src/business-report/business-report.controller.internal.ts]: Is the `where` necessary? It counts the number of existing reports for a specific business, not the number of BusinessReports the customer has

File: services/workflows-service/src/business-report/business-report.controller.internal.ts [hunk 4]
```diff
@@ -55,6 +56,26 @@ export class BusinessReportControllerInternal {
...
+            businessId: businessCorrelationId,
+          },
+        },
+        [currentProjectId],
+      );
+
+      if (businessReportsCount >= maxBusinessReports) {
+        throw new BadRequestException(
```
Comment (tomer-shvadron) [inline services/workflows-service/src/business-report/business-report.controller.internal.ts:66]: See my previous comment about HTTP error code 429

**Review by tomer-shvadron — APPROVED** [review]

File: services/workflows-service/src/business-report/business-report.controller.internal.ts (hunk 3 above)
Comment (MatanYadaev) [reply services/workflows-service/src/business-report/business-report.controller.internal.ts]: 🤦
Thanks for the catch
Copilot autocompleted it and I didn't review it

File: apps/backoffice-v2/src/domains/business-reports/hooks/mutations/useCreateBusinessReportMutation/useCreateBusinessReportMutation.tsx (hunk 2 above)
Comment (MatanYadaev) [reply apps/backoffice-v2/src/domains/business-reports/hooks/mutations/useCreateBusinessReportMutation/useCreateBusinessReportMutation.tsx:55]: It's not related... 429 is for rate limit errors.

Comment (tomer-shvadron) [reply apps/backoffice-v2/src/domains/business-reports/hooks/mutations/useCreateBusinessReportMutation/useCreateBusinessReportMutation.tsx:55]: From what I understand, you could send the reason for the "too many requests" in the `message` property of the error. I'm guessing 400 or 403 serve the same purpose in that regard, only 400 is more "generic". Anyway 400 is fine

**Pull request merged**